server/otlp/ingester.go

```diff
@@ -55,69 +83,54 @@ func (i ingester) Ingest(ctx context.Context, request *pb.ExportTraceServiceRequ
 	}, nil
 }
 
-func (i ingester) getSpansByTrace(request *pb.ExportTraceServiceRequest) map[trace.TraceID][]model.Span {
-	otelSpans := make([]*v1.Span, 0)
-	for _, resourceSpan := range request.ResourceSpans {
-		for _, spans := range resourceSpan.ScopeSpans {
-			otelSpans = append(otelSpans, spans.Spans...)
+func (i ingester) traces(input []*v1.ResourceSpans) []traces.Trace {
+	spansByTrace := map[string][]*v1.Span{}
+
+	for _, rs := range input {
+		for _, il := range rs.ScopeSpans {
+			for _, span := range il.Spans {
+				traceID := trace.TraceID(span.TraceId).String()
+				i.log("adding span %s to trace %s", hex.EncodeToString(span.SpanId), traceID)
+				spansByTrace[traceID] = append(spansByTrace[traceID], span)
+			}
 		}
 	}
 
-	spansByTrace := make(map[trace.TraceID][]model.Span)
+	i.log("sorted %d traces", len(spansByTrace))
 
-	for _, span := range otelSpans {
-		traceID := traces.CreateTraceID(span.TraceId)
-		var existingArray []model.Span
-		if spansArray, ok := spansByTrace[traceID]; ok {
-			existingArray = spansArray
-		} else {
-			existingArray = make([]model.Span, 0)
-		}
-
-		existingArray = append(existingArray, *traces.ConvertOtelSpanIntoSpan(span))
-		spansByTrace[traceID] = existingArray
+	modelTraces := make([]traces.Trace, 0, len(spansByTrace))
+	for traceID, spans := range spansByTrace {
+		i.log("creating trace %s with %d spans", traceID, len(spans))
+		modelTraces = append(
+			modelTraces,
+			traces.FromSpanList(spans),
+		)
 	}
 
-	return spansByTrace
+	return modelTraces
 }
 
-func (e ingester) saveSpansIntoTest(ctx context.Context, traceID trace.TraceID, spans []model.Span, requestType string) error {
-	run, err := e.runRepository.GetRunByTraceID(ctx, traceID)
-	if err != nil && strings.Contains(err.Error(), "record not found") {
-		// span is not part of any known test run. So it will be ignored
+func (i ingester) notify(ctx context.Context, trace traces.Trace, requestType string) error {
+	run, err := i.runGetter.GetRunByTraceID(ctx, trace.ID)
+	if errors.Is(err, sql.ErrNoRows) {
+		// trace is not part of any known test run, no need to notify
 		return nil
 	}
-
 	if err != nil {
-		return fmt.Errorf("could not find test run with traceID %s: %w", traceID.String(), err)
-	}
-
-	if run.State != test.RunStateAwaitingTrace {
-		// test is not waiting for trace, so we can completely ignore those as they might
```

Comment:
test integrity is now safe, since the results are saved to another table. the ingestor has no impact on the runs table